### PR TITLE
New request model

### PR DESCRIPTION
**Pull request by scottinet**
Refactor of the Kuzzle request model, to make it clearer and cleaner.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,112 +130,185 @@ var requestObject = new RequestObject({
...
 |`controller` |`string`| Kuzzle controller handling the action to perform |
-|`data._id`|`string`| Document unique identifier |
-|`data.body`|`object`| Contains response data (document content, search results, ...) |
-|`error`|`Error object`| In case of error, contains the error object |
 |`index` |`string`| Data index |
-|`metadata`|`object`| Contains request specific metadata |
-|`status`|`integer`| Request status, using [HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes) |
-|`timestamp`|`integer`| Request timestamp |
+|`metadata`|`object`| Client's request specific metadata |
+|`token`|`string`| JWT Authentication token |
```
Comment (ballinette) [inline README.md]: why do we need token in both Request.input AND Request.context

File: README.md [hunk 2]
```diff
@@ -89,112 +130,185 @@ var requestObject = new RequestObject({
...
 
-#### Example
 
-```js
-var ResponseObject = require('kuzzle-common-objects').Models.responseObject;
+## `errors.KuzzleError`
 
-var responseObject = new ResponseObject(requestObject, {response: 'data'});
-```
+Inherits from `Error`. Base class inherited by Kuzzle error objects.
```
Comment (ballinette) [inline README.md]: So, KuzzleError is like an **Abstract Class**

File: README.md (hunk 1 above)
Comment (ballinette) [inline README.md]: OK, forget it, it's not the same one : 
* in `Request.input` : the JWT token string coming from request headers
* in `Request.context`: the token object (decrypted from the input's jwt) used internally within Kuzzle kernel.

It's quite disturbing, while reading the code, to see 2 different representations of `token` attributes...

Maybe we could rename Request input's `token` attribute to `jwt` ?

File: README.md (hunk 2 above)
Comment (scottinet) [inline README.md]: Not sure the concept can be applied to javascript, but yes.

File: README.md (hunk 1 above)
Comment (scottinet) [inline README.md]: Fine by me.

File: lib/models/requestInput.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    Object.seal(this);
+
+    Object.keys(data).forEach(k => {
+      if (!this.hasOwnProperty(k) && !this.resource.hasOwnProperty(k)) {
+        this.args[k] = data[k];
+      }
+    });
+
+    this.jwt = data.jwt;
+    this.metadata = assert.assertObject('metadata', data.metadata);
```
Comment (ballinette) [inline lib/models/requestInput.js]: Why still check assertObject() here ?
Isn't it already done by the setter ?

Comment (scottinet) [inline lib/models/requestInput.js]: Dang! Fixed.

File: lib/models/requestContext.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
+'use strict';
+
+const assert = require('../utils/assertType');
+
+// Private variables
+let
+  _connectionId = new WeakMap(),
+  _protocol = new WeakMap(),
+  _token = new WeakMap(),
+  _user = new WeakMap();
```
Comment (benoitvidis) [inline lib/models/requestContext.js]: Shouldn't these be defined in the constructor instead?
Defining these vars in the module scope make them behave like static properties to me.

File: lib/models/requestInput.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
+'use strict';
+
+const
+  InternalError = require('../errors/internalError'),
+  assert = require('../utils/assertType');
+
+// Private variables
+let
+  _jwt = new WeakMap(),
```
Comment (benoitvidis) [inline lib/models/requestInput.js]: Same as above.

**Comment by benoitvidis** [comment]
Sorry, my bad, I had not seen the `this` used as key for the weakmap. LGTM

**Pull request merged**